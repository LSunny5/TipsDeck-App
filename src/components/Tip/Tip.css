.TipContainer {
    width: 85%;
    margin: 20px auto;

}

.ratingAverage {
    margin: 10px;
}

.rateMessage {
    margin-top:10px;
    font-weight: bolder;
}

.starBox {
    margin: 0 10px 10px;
}

.tipTitle {
    font-size: 2em;
}

.catName {
    font-size: 1.5em;
}

.tipText {
    text-align: left;
    margin: 10px;
}

.tipDirections {
    text-align: left;
    margin: 20px;
}

.lefttext {
    text-align: left;
}

.bold {
    font-weight: bold;
}

.sourceBox {
    margin: 20px;
}

.wordBreak {
    overflow-wrap: break-word;
    word-wrap: break-word;
}

.buttonBox {
    width: 80%;
    margin: 0 auto;
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    align-items: center;
}

.deleteTipButton {
    background: lightcoral;
    border: 1px solid black;
    box-shadow: none;
    padding: 13px;
    width: 75px;
}

.editTipButton {
    background: white;
    text-decoration: none;
    color: black;
    border: 1px solid black;
    padding: 10px;
    width: 80px;
}