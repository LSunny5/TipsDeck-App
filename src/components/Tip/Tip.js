import React from 'react';
import './Tip.css';
import TipsDeckContext from '../../TipsDeckContext';
import { findTip, getCategoryName } from '../../TipsDeckHelpers';
import { NavLink } from 'react-router-dom'
import Rating from '../Rating/Rating';
import PropTypes from 'prop-types';
import config from '../../config';

class Tip extends React.Component {
    static contextType = TipsDeckContext;


    handleClickDelete = (cName) => {
        const { id } = this.props.match.params;

        //fetch code here for note
        fetch(`${config.APIEndpoint}/Tips/${id}`, {
            method: 'DELETE',
            headers: {
                'content-type': 'application/json'
            },
        })
            .then(response => {
                if (!response.ok)
                    return response.json().then(error => Promise.reject(error))
                return response
            })
            .then(() => {
                this.context.deleteTip(id);
                this.props.history.push(`/Category/${cName}`);
                window.location.reload();
            })
            .catch(error => {
                alert('Could not delete note:  ' + error);
            })
    }

    handleClickCancel = () => {
        alert('Tip was not deleted');
    }

    render() {
        const { tips = [], categories = [] } = this.context;
        const { id } = this.props.match.params;
        const targetTip = findTip(tips, id) || {};
        const cat = getCategoryName(categories, targetTip.category_id) || {};

        return (
            <section className="TipContainer">
                <h1 className="tipTitle">{targetTip.tipname}</h1>
                <h2 className="catName">{cat.category}</h2>

                <Rating value={targetTip.rating} />

                <div className="tipText">
                    <p className="bold lefttext">Description: </p>
                    {targetTip.tipdescription}
                </div>
                <div className="tipText">
                    <p className="bold lefttext">Directions: </p>
                    {targetTip.directions}
                </div>
                <div className="tipText sourceBox">
                    <p className="bold lefttext">Source: </p>
                    <a target='_blank' rel="noopener noreferrer" href={targetTip.sourceurl}>
                        {targetTip.sourcetitle}
                    </a>
                </div>

                <div className="buttonBox">
                    <NavLink
                        className="editTipButton"
                        to={`/Category/${cat.category}/${id}/editTip`}
                    >
                        Edit
                    </NavLink>

                    <button
                        className="deleteTipButton"
                        type='button'
                        onClick={() => {
                            if (window.confirm('Are you sure you wish to delete the tip?') ?
                                this.handleClickDelete(cat.category) : this.handleClickCancel())
                                this.handleClickCancel()
                        }}
                        aria-pressed='false'
                        aria-label='delete tip button'
                    >
                        Delete
                    </button>
                </div>
            </section>
        );
    }
}

Tip.propTypes = {
    categories: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
            category: PropTypes.string.isRequired,
        })
    ),
    tips: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
            category_id: PropTypes.number.isRequired,
            tipname: PropTypes.string.isRequired,
            tipdescription: PropTypes.string.isRequired,
            directions: PropTypes.string.isRequired,
            sourcetitle: PropTypes.string,
            sourceurl: PropTypes.string,
            rating: PropTypes.number.isRequired,
            numraters: PropTypes.number.isRequired,
        })
    ),
    id: PropTypes.number,
};

export default Tip;