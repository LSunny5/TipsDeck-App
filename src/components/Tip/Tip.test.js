import React from 'react';
import ReactDOM from 'react-dom';
import Tip from './Tip';
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'
import { BrowserRouter } from 'react-router-dom';

describe(`Tip component`, () => {
    const props = {
        match: {
          params: {
            id: 3,
          }
        }
    }

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <BrowserRouter>
                <Tip {...props} />
            </BrowserRouter>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the Tip form', () => {
        const wrapper = shallow(<Tip {...props} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })
});