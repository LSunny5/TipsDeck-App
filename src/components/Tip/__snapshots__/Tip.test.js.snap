// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tip component renders the Tip form 1`] = `
<section
  className="TipContainer"
>
  <div
    className="tipBackground"
  >
    <h1
      className="tipTitle"
    />
    <h2
      className="catName"
    />
    <Rating />
    <div
      className="tipText"
    >
      <p
        className="bold lefttext"
      >
        Description: 
      </p>
      <p
        className="indent"
      />
    </div>
    <div
      className="tipText"
    >
      <p
        className="bold lefttext"
      >
        Directions: 
      </p>
      <p
        className="indent"
      />
    </div>
    <div
      className="tipText sourceBox"
    >
      <p
        className="bold lefttext"
      >
        Source: 
      </p>
      <a
        rel="noopener noreferrer"
        target="_blank"
      />
    </div>
    <div
      className="buttonBox"
    >
      <NavLink
        className="editTipButton button"
        to="/Category/undefined/3/editTip"
      >
        Edit
      </NavLink>
      <button
        aria-label="delete tip button"
        aria-pressed="false"
        className="deleteTipButton button"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</section>
`;
